### ISSUE TITLE
structs with public members get extra pub prefix

### ISSUE THREAD

**Issue by jonhoo** [nrc/rustfmt#95]
rustfmt adds an extra `pub` modifier to public fields in structs, causing the code to no longer compile.

```rust
// before rustfmt
struct X {
    pub x : u64,
}

// after rustfmt
struct X {
    pub    pub x: u64,
}
```

**Comment by chellmuth** [comment]
I think the root cause is a bug in rustc. It looks like the span for `StructField` starts at the field name instead of the pub keyword.

After I applied this patch to `libsyntax/parse/parser.rs` in rustc, rustfmt handled public struct fields correctly:

```
diff --git a/src/libsyntax/parse/parser.rs b/src/libsyntax/parse/parser.rs
--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
     /// Parse a structure field
     fn parse_name_and_ty(&mut self, pr: Visibility,
                          attrs: Vec<Attribute> ) -> PResult<StructField> {
-        let lo = self.span.lo;
+        let lo = match pr {
+            Inherited => { self.span.lo }
+            Public => { self.last_span.lo }
+        };
         if !self.token.is_plain_ident() {
             return Err(self.fatal("expected ident"));
         }
         let name = try!(self.parse_ident());
         try!(self.expect(&token::Colon));
         let ty = try!(self.parse_ty_sum());
         Ok(spanned(lo, self.last_span.hi, ast::StructField_ {
             kind: NamedField(name, pr),
             id: ast::DUMMY_NODE_ID,
             ty: ty,
             attrs: attrs,
         }))
     }
```
@nrc: Can you confirm this is a rustc bug and not intentional behavior? It seems like a bug to me, but I'm new to compiler internals. I can open a pull request to get this change applied.

**Comment by nrc** [comment]
@chellmuth nice find! Yeah, rustc looks wrong here. Please could you open a PR against Rust?

**Comment by chellmuth** [comment]
Issue: https://github.com/rust-lang/rust/issues/26083
PR: https://github.com/rust-lang/rust/pull/26084

**Comment by chellmuth** [comment]
This is fixed on the latest nightly, #100 adds a regression test

**Comment by nrc** [comment]
Fixed by #100
